pkg/local_object_storage/peapod/peapod.go

```diff
@@ -0,0 +1,174 @@
+package peapod
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sync"
+	"time"
+
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.etcd.io/bbolt"
+)
+
+type batch struct {
+	initErr error
+
+	tx *bbolt.Tx
+
+	nonIdle bool
+
+	commitErr   error
+	chCommitted chan struct{}
+
+	bktRootMtx sync.Mutex
+	bktRoot    *bbolt.Bucket
+}
+
+// Peapod provides storage for relatively small NeoFS binary object (peas).
+// Peapod is a single low-level key/value database optimized to work with big
+// number of stored units.
+type Peapod struct {
+	bolt *bbolt.DB
+
+	currentBatchMtx sync.RWMutex
+	currentBatch    *batch
+
+	chClose     chan struct{}
+	chFlushDone chan struct{}
+}
+
+var rootBucket = []byte("root")
+
+// New returns initialized Peapod instance located at the given path. Resulting Peapod
+// must be finally closed.
+func New(path string, readOnly bool) (*Peapod, error) {
+	db, err := bbolt.Open(path, 0600, &bbolt.Options{
+		ReadOnly: readOnly,
+		Timeout:  100 * time.Millisecond, // to handle flock
+	})
+	if err != nil {
+		return nil, fmt.Errorf("open BoltDB instance: %w", err)
+	}
+
+	if !readOnly {
+		err = db.Update(func(tx *bbolt.Tx) error {
+			_, err = tx.CreateBucketIfNotExists(rootBucket)
+			return err
+		})
+		if err != nil {
+			return nil, fmt.Errorf("create root bucket in BoltDB instance: %w", err)
+		}
+	}
+
+	res := &Peapod{
+		bolt:        db,
+		chClose:     make(chan struct{}),
+		chFlushDone: make(chan struct{}),
+	}
+
+	res.beginNewBatch()
+
+	go res.flushLoop()
+
+	return res, nil
+}
+
+// Close syncs data and closes the database.
+func (x *Peapod) Close() error {
+	close(x.chClose)
+	<-x.chFlushDone
+	return x.bolt.Close()
+}
+
+func (x *Peapod) flushLoop() {
+	defer close(x.chFlushDone)
+
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-x.chClose:
+			// commit current transaction to prevent bbolt.DB.Close blocking
+			x.flushCurrentBatch(false)
+			return
+		case <-ticker.C:
```

Comment:
do we want to finish batches earlier if there are "too many" operations?

Reply:
in current model: yes cuz we don't wanna block earlier operations due to load. Like i mentioned here https://github.com/nspcc-dev/neofs-node/pull/2462#discussion_r1278913801 external load and flush performance isn't analyzed for now (benchmarks show perf improvement even with this simple model). 

we'll leave an optimization issue to test peapod against itself